### PR TITLE
rework common array operations into functions that can be migrated to the GC

### PR DESCRIPTION
**Pull request by schveiguy**
The entire purpose of this PR is to rework the array code in lifetime.d to use a small set of functions to avoid having intimate knowledge of the array metdata details leaking into lifetime.d

In a future PR these mechanisms will be moved into the GC, and lifetime will implement its features by calling these functions on the GC directly.

This was split into 2 commits. It may need to be split further. It might even need to be split into multiple PRs.

This also separates out the setting of the finalizer with setting array length. I did not realize before this PR that the finalizer was always being set when the length was set. I'm also removing some dependence on TypeInfo there. In retrospect, maybe this part should be moved to a later PR.

There is one unittest I commented out, because it depends on array appending forwarding current bits to the expanded array instead of using the expected type bits from the type directly. I'm not entirely sure if we want to keep this piece. I want to see how buildkite does with this.

### CODE REVIEW

File: druntime/src/rt/lifetime.d [hunk 1]
```diff
@@ -32,6 +32,161 @@ private
...
+    else if (!atomic)
+        __insertBlkInfoCache(info, null);
+
+    // this should always work.
+    return __setArrayAllocLengthImpl(info, newUsed, atomic, existingUsed, typeInfoSize);
+}
+
+private bool shrinkArrayUsed(void[] slice, size_t existingUsed, bool atomic = false) nothrow
+{
+    if(existingUsed < slice.length)
```
Comment (thewilsonator) [inline druntime/src/rt/lifetime.d:110]: space between `if` and `(`

File: druntime/src/rt/lifetime.d [hunk 2]
```diff
@@ -1893,7 +1762,13 @@ unittest
     assert(info2.base is carr2.ptr); // no offset, the capacity is small.
 }
 
-unittest
+// I think this whole test is invalid, as is the bug report. Setting internal
```
Comment (thewilsonator) [inline druntime/src/rt/lifetime.d:1765]: given this, I think this PR (in conjunction with later PRs, you need not do this now) should have a changelog entry and this should be made deprecated.

**Review by thewilsonator — APPROVED** [review]
Some minor nits. CI failure is unrelated.

**Comment by schveiguy** [comment]
OK, I think I'm going to support the unittest for this. It's an extra query needed to get the existing bits to make sure things conform to the status quo. But the general path forward is to deprecate random bit setting. As long as random bit setting is supported, then the behavior "makes sense".

The better way is to get all this stuff merged into the GC, and then address how to modify the GC to migrate away from unsanctioned bit setting.

File: druntime/src/rt/lifetime.d (hunk 1 above)
Comment (schveiguy) [reply druntime/src/rt/lifetime.d:110]: Surprised the style checker passed this?

File: druntime/src/rt/lifetime.d (hunk 2 above)
Comment (schveiguy) [reply druntime/src/rt/lifetime.d]: moot now I think. Should be the same behavior.

**Review by thewilsonator — APPROVED** [review]

**Comment by CyberShadow** [comment]
This pull request introduced a regression: https://github.com/dlang/dmd/issues/21826

**Pull request merged**